Enonic UI: Externalize Projects Data Management #9285

reisfmb opened the pull request:
## 1 - Some comments about the Acceptance criteria on #9285 

### 1.1 - No duplicate project loading requests across application

There are many places requesting projects directly (see `new ProjectListRequest`), and manipulating the active one via `ProjectContext`.

From my understanding it was out of the scope of this task to replace all `ProjectListRequest` and `ProjectContext` usage with the new projects store, but rather create the store and start using it in minimal context, and that's what I've done on `ProjectSelectionDialog`.

Note that I still left the `ProjectContext `manipulation there, with proper code comments to remember us to get rid of it once we fully migrate to use only the projects store.

### 1.2 - Project loading logic removed from ContentAppBar

I didn't touch `ContentAppBar` because we're not instantiating it anymore after the `AppBarElement` replacement of `ContentAppBar` (task #9279)

The only place still calling it is the `SettingsAppPanel` constructor, and I already added a comment there for future removal, but that depends on updating lib-admin-ui as well.

---

## 2 - Some extra comments and questions (reason for PR being draft)
a. Currently we only automatically set the active project if there's no project set as active and there's only one project loaded in the store. If none is active and there is more than one project, then we don't automayically set the active project! What should be the logic in this case?

b. I could've gone further and improved the `ProjectSelectionDialog` and `main.ts`, removing `ProjectListRequest` usage from both, and instead, getting projects from the store... I have two reasons for not doing that yet:
1) because I was unsure if that would be out of the scope
2) because there would be fricction with the `readonly` projects from the projects store (since most functions expect projects as `Project[]` not `Readonly<Project>[]`)

c. What should happen in case no project is selected?!

reisfmb committed:
Enonic UI: Externalize Projects Data Management #9285

Created projects store
Added minor adjustments to sidebarWidgets store
Used projects store methods in ProjectSelectionDialog
Used active project name from projects store in AppBar
Added useStore on AppBar

edloidas commented:
**1.1**
Okay, if it’s too much to remove, keep it as is. Just make sure we don’t make excessive requests when using the stores themselves.

**1.2**
So you’re suggesting removing `ContentAppBar` altogether rather than removing its functionality piece by piece? I’m not sure that’s a better approach, but if it’s too much work otherwise and your way is easier, then okay.

**2.a**
We show the dialog so that the user can select a project. It will be implemented within the dialog, e.g.

```jsx
<ProjectDialog open={activeProjectId == null && projects.length > 1} />
```

**2.b**
Not in scope. Ignore it. The project selection dialog is a separate task.

**3.c**
If there are no projects, we don’t select anything — just disable the project selection button and change its label.
If there’s exactly one project and no selection, we automatically select it after loading.
If there are two or more projects, we show the dialog so the user can make a selection manually.

reisfmb commented:
1.1
Yep, its quite a lot. On the store we only load once and then only at Project create/update events. On delete we manually delete the deleted project from the store (no need to reload).

1.2
Exactly, because after `AppBar` implementation we basically don't use `ContentAppBar` anymore.

That raises an UX concern: what should happen to the `AppBar` when we are outside of the content widget? We're essentially breaking the current behavior. Also the "tabs" are gone to (as we internally discussed).

2.(a,b,c) 
👍 

---

Tomorrow morning I'll fix conflicts and change from draft to ready!

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/layout/AppShell/AppBar.tsx:
```diff
@@ -19,7 +23,7 @@ const AppBar = (): ReactElement => {
                 aria-label={useI18n('wcag.appbar.project.label')}
             >
-                Default Project
+                {$activeProjectName.get()}
```
Better use `label` property of our Button component. More consistent.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/sidebarWidgets.store.ts:
```diff
@@ -17,6 +17,8 @@ export const $sidebarWidgets = map<WidgetsStore>({
 });
 
 export function setActiveWidget(widget: Readonly<Widget> | undefined): void {
+    const existsInStore = $sidebarWidgets.get().widgets.some((p) => getWidgetKey(p) === getWidgetKey(widget));
```
Just a sm<i>aaa</i>ll change here: start with `is` for consistency with our code style. We drop `is`, `has`, etc only in prop names.

edloidas commented on modules/lib/src/main/resources/assets/js/app/dialog/ProjectSelectionDialog.ts:
```diff
@@ -47,6 +48,8 @@ export class ProjectSelectionDialog
             itemView.onClicked((event: MouseEvent) => {
                 if (!event.ctrlKey && !event.shiftKey) {
                     if (itemView.isSelectable()) {
+                        setActiveProject(itemView.getProject());
```
Please, add annotation to the `ProjectContext` file itself.
```ts
/**
 * @deprecated The class due to be replaced by the projects.store
 */
export class ProjectContext {
```

edloidas reviewed (commented):
And please,

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/layout/AppShell/AppBar.tsx:
```diff
@@ -6,8 +6,12 @@ import {ProjectSelectionDialog} from '../../../../app/dialog/ProjectSelectionDia
+import {useStore} from '@nanostores/preact';
 
 const AppBar = (): ReactElement => {
+    useStore($activeProjectName);
```
Not technically correct. You should do
```ts
const activeProjectName = useStore($activeProjectName);
// ...
label={activeProjectName}
```

reisfmb replied:
Yep, sure... Idk why I did it that way. Probably was testing something and forget to change to the proper way before opening the PR. Thx.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/projects.store.ts:
```diff
@@ -0,0 +1,104 @@
+    activeProjectId: undefined,
+});
+
+export function setActiveProject(project: Readonly<Project> | undefined): void {
+    const isProjectInStore = $projects.get().projects.some((p) => getProjectKey(p) === getProjectKey(project));
+    if (!isProjectInStore) throw new Error('Project not found in store');
+    $projects.setKey('activeProjectId', getProjectKey(project));
+}
```
Put under the stores

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/projects.store.ts:
```diff
@@ -0,0 +1,104 @@
+//
+export function getProjectKey(project: Readonly<Project> | undefined): string | undefined {
+    return project?.getName();
+}
```
Maybe it's `getProjectId` if it's used for setting ID?

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/sidebarWidgets.store.ts:
```diff
@@ -17,6 +17,8 @@ export const $sidebarWidgets = map<WidgetsStore>({
 
 export function setActiveWidget(widget: Readonly<Widget> | undefined): void {
+    const existsInStore = $sidebarWidgets.get().widgets.some((p) => getWidgetKey(p) === getWidgetKey(widget));
+    if (!existsInStore) throw new Error('Widget not found in store');
```
It's not really correct to throw an error here, because you are not handling it in the code later. And it's not something to be handled here. Just return.

edloidas commented on modules/lib/src/main/resources/assets/js/v6/features/store/projects.store.ts:
```diff
@@ -0,0 +1,104 @@
+
+export function setActiveProject(project: Readonly<Project> | undefined): void {
+    const isProjectInStore = $projects.get().projects.some((p) => getProjectKey(p) === getProjectKey(project));
+    if (!isProjectInStore) throw new Error('Project not found in store');
```
Just return. No need to throw error. You expect to (and actually do) pass a valid project, that you took from `$projects`. There should not be any error there.

reisfmb committed:
Enonic UI: Externalize Projects Data Management #9285

Removed unecessary error throw on projects and widgets stores

Added other minor improvements to projects store
